src/compiler/core.ts

```diff
@@ -2458,6 +2458,11 @@ namespace ts {
             throw e;
         }
 
+        export function assertNever(_member: never, message?: string, stackCrawlMark?: Function): never {
+            fail(message, stackCrawlMark || assertNever);
+            throw "Unreachable";

```

Comment:
You could make `fail` return `never` and write `return fail(message, stackCrawlMark || assertNever)`

Reply:
It's nice to check that the type of the member is `never` though.

Reply:
I think you just mean that `throw "Unreachable";` is not required if `fail` just returns `never` - which seems fine to me; so I've done that.